driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +184,356 @@ public int getMaxWireVersion() {
+                lock.unlock();
+            }
+        }
 
-    private void handleGetMoreQueryResult(final AsyncConnection connection, final SingleResultCallback<List<T>> callback,
-                                          final CommandCursorResult<T> result) {
-        logQueryResult(result);
-        cursor.set(result.getServerCursor());
-        if (isClosePending) {
+        /**
+         * Thread-safe.
+         */
+        private void endOperation() {
+            boolean doClose = false;
+            lock.lock();
             try {
-                connection.release();
-                if (result.getServerCursor() == null) {
-                    connectionSource.release();
+                State localState = state;
+                if (localState == State.OPERATION_IN_PROGRESS) {
+                    state = State.IDLE;
+                } else if (localState == State.CLOSE_PENDING) {
+                    state = State.CLOSED;
+                    doClose = true;
+                } else if (localState != State.CLOSED) {
+                    fail(localState.toString());
                 }
-                endOperationInProgress();
             } finally {
-                callback.onResult(null, null);
+                lock.unlock();
             }
-        } else if (result.getResults().isEmpty() && result.getServerCursor() != null) {
-            getMore(connection, assertNotNull(result.getServerCursor()), callback);
-        } else {
-            if (limitReached()) {
-                killCursor(connection);
-                connection.release();
-            } else {
-                connection.release();
-                if (result.getServerCursor() == null) {
-                    connectionSource.release();
+            if (doClose) {
+                doClose();
+            }
+        }
+
+        /**
+         * Thread-safe.
+         */
+        void close() {
+            boolean doClose = false;
+            lock.lock();
+            try {
+                State localState = state;
+                if (localState == State.OPERATION_IN_PROGRESS) {
+                    state = State.CLOSE_PENDING;
+                } else if (localState != State.CLOSED) {
+                    state = State.CLOSED;
+                    doClose = true;
                 }
+            } finally {
+                lock.unlock();
+            }
+            if (doClose) {
+                doClose();
+            }
+        }
+
+        /**
+         * This method is never executed concurrently with either itself or other operations
+         * demarcated by {@link #tryStartOperation()}/{@link #endOperation()}.
+         */
+        private void doClose() {
+            if (skipReleasingServerResourcesOnClose) {
+                serverCursor = null;
             }
-            endOperationInProgress();
 
-            if (result.getResults().isEmpty()) {
-                callback.onResult(null, null);
+            if (getServerCursor() != null) {
```

Comment:
This commit was split out from a previous PR as this isn't the end state of the code I'd rather not make changes here just for them to be refactored out in #401 

Could you review the code there instead?

Reply:
I believe all the points are addressed in #401

> I vaguely remember that synchronous doClose couldn't use executeWithConnection for some reason.

I can't determine a reason why that would be the case, it would have to have a connection to kill cursors if a user calls `cursor.close()`.

Reply:
nm - it uses `connection()` which doesn't modify any exceptions. Will update in #401.